// @slack/client/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var logger_1 = require("@slack/logger");
exports.LogLevel = logger_1.LogLevel;
// Using a wildcard (*) export because listing each individual method is cumbersome and hard to keep up to date.
// The following export contains:
// addAppMetadata,
// TLSOptions,
// retryPolicies,
// RetryOptions,
// WebClient,
// WebClientOptions,
// WebAPICallOptions,
// WebAPICallResult,
// WebAPIPlatformError,
// WebAPIRequestError,
// WebAPIHTTPError,
// WebAPIRateLimitedError,
// WebAPICallError,
// WebClientEvent,
// and all method definitions
__export(require("@slack/web-api"));
var web_api_1 = require("@slack/web-api");
exports.WebAPIErrorCode = web_api_1.ErrorCode;
var rtm_api_1 = require("@slack/rtm-api");
exports.RTMClient = rtm_api_1.RTMClient;
var rtm_api_2 = require("@slack/rtm-api");
exports.RTMAPIErrorCode = rtm_api_2.ErrorCode;
var webhook_1 = require("@slack/webhook");
exports.IncomingWebhook = webhook_1.IncomingWebhook;
var webhook_2 = require("@slack/webhook");
exports.WebhookErrorCode = webhook_2.ErrorCode;
// Merge the two ErrorCode constants into one, to make this as similar to @slack/client@4 as possible
const web_api_2 = require("@slack/web-api");
const rtm_api_3 = require("@slack/rtm-api");
const webhook_3 = require("@slack/webhook");
// tslint:disable-next-line:variable-name
exports.ErrorCode = Object.assign({}, web_api_2.ErrorCode, rtm_api_3.ErrorCode, webhook_3.ErrorCode);
